File: baseunits/modules/MangaChanRU.pas
unit MangaChanRU;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, Math, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread, synautil, RegExpr;

implementation

const
  dirurl = '/manga/new';
  perpage = 20;

function GetDirectoryPageNumber(const MangaInfo: TMangaInformation;
  var Page: Integer; const WorkPtr: Integer; const Module: TModuleContainer): Integer;
var
  s: String;
begin
  Result := NET_PROBLEM;
  Page := 1;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  if MangaInfo.FHTTP.GET(Module.RootURL + dirurl) then begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        s := XPathString('//*[@id="pagination"]/a[last()]/@href');
        if s <> '' then Page := StrToIntDef(SeparateRight(s, 'offset='), 1);
        if Page > 1 then
          Page := ceil(Page / perpage) + 1;
      finally
        Free;
      end;
  end;
end;

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  v: IXQValue;
  s: String;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  s := Module.RootURL + dirurl;
  if AURL <> '0' then s += '?offset=' + IntToStr(StrToInt(AURL) * perpage);
  if MangaInfo.FHTTP.GET(s) then begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        for v in XPath('//*[@class="content_row"]//a[@class="title_link"]') do begin
          ALinks.Add(v.toNode.getAttribute('href'));
          ANames.Add(v.toString);
        end;
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  v: IXQValue;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.FHTTP, MangaInfo.mangaInfo do
  begin
    if GET(FillHost(Module.RootURL, AURL)) then
    begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do
        try
          coverLink := MaybeFillHost(Module.RootURL, XPathString('//img[@id="cover"]/@src'));
          if title = '' then title := XPathString('//*[@class="name_row"]/h1');
          authors := XPathString('//*[@class="item" and contains(.,"Автор")]/following-sibling::*[1]');
          genres := XPathString('//*[@class="item" and contains(.,"Тэги")]/following-sibling::*[1]');
          status := MangaInfoStatusIfPos(
            XPathString('//*[@class="item" and contains(.,"Загружено")]/following-sibling::*[1]'),
            'продолжается',
            '');
          summary := XPathString('//*[@id="description"]/text()');
          for v in XPath('//*[@class="table_cha"]//*[@class="manga"]/a') do begin
            chapterLinks.Add(v.toNode.getAttribute('href'));
            chapterName.Add(v.toString);
          end;
          if chapterLinks.Count = 0 then
            for v in XPath('//a[.="Читать онлайн"]') do
            begin
              chapterLinks.Add(v.toNode.getAttribute('href'));
              chapterName.Add(title);
            end;
          InvertStrings([chapterLinks, chapterName]);
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  Source: TStringList;
  i, j: Integer;
  s: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.FHTTP, DownloadThread.Task.Container do begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(FillHost(Module.RootURL, AURL)) then begin
      Result := True;
      Source := TStringList.Create;
      try
        Source.LoadFromStream(Document);
        if Source.Count > 0 then
          for i := 0 to Source.Count - 1 do
            if Pos('"fullimg":', Source[i]) > 0 then begin
              s := SeparateRight(Source[i], ':');
              s := TrimChar(s, ['[', ']', ',']);
              PageLinks.CommaText := s;
              if PageLinks.Count > 0 then
                with TRegExpr.Create('(?i)//im(\d*\.)') do
                  try
                    for j := 0 to PageLinks.Count - 1 do
                      PageLinks[j] := Replace(PageLinks[j], '//img$1', True);
                  finally
                    Free;
                  end;
              Break;
            end;
      finally
        Source.Free;
      end;
    end;
  end;
end;

procedure RegisterModule;

  procedure AddWebsiteModule(const AWebsite, ARootURL, ACategory: String);
  begin
    with AddModule do
    begin
      Website := AWebsite;
      RootURL := ARootURL;
      Category := ACategory;
      SortedList := True;
      OnGetDirectoryPageNumber := @GetDirectoryPageNumber;
      OnGetNameAndLink := @GetNameAndLink;
      OnGetInfo := @GetInfo;
      OnGetPageNumber := @GetPageNumber;
    end;
  end;

begin
  AddWebsiteModule('MangaChanRU', 'http://mangachan.me', 'Russian');
  AddWebsiteModule('HentaiChanRU', 'http://hentai-chan.me', 'H-Sites');
  AddWebsiteModule('YaoiChanRU', 'http://yaoichan.me', 'H-Sites');
end;

initialization
  RegisterModule;

end.

File: lua/modules/MangaChanRu.lua
local dirurl = '/manga/new';
local perpage = 20;

function getinfo()
  mangainfo.url=MaybeFillHost(module.RootURL, url)
  if http.get(mangainfo.url) then
    x=TXQuery.Create(http.document)
	  mangainfo.coverlink=MaybeFillHost(module.RootURL, x.xpathstring('//img[@id="cover"]/@src'))
	  mangainfo.title=x.xpathstring('//*[@class="name_row"]/h1')
    mangainfo.authors=x.xpathstring('//*[@class="item" and contains(.,"Автор")]/following-sibling::*[1]')
    mangainfo.genres=x.xpathstring('//*[@class="item" and contains(.,"Тэги")]/following-sibling::*[1]')
    mangainfo.status=MangaInfoStatusIfPos(x.xpathstring('//*[@class="item" and contains(.,"Загружено")]/following-sibling::*[1]'), 'продолжается', '')
    mangainfo.summary=x.xpathstringall('//*[@id="description"]/text()')
    if module.website=='MangaChanRU' then x.xpathhrefall('//table[@class="table_cha"]//div[@class="manga2"]/a', mangainfo.chapterlinks, mangainfo.chapternames) end
    if module.website=='YaoiChanRU' then x.xpathhrefall('//table[@class="table_cha"]//div[@class="manga"]/a', mangainfo.chapterlinks, mangainfo.chapternames) end
    if module.website=='HentaiChanRU' then
      local v=x.xpath('//div[@id="manga_images"]/a')
      for i = 1, v.count do
        v1 = v.get(i)
        mangainfo.chapterlinks.add(v1.getattribute('href'))
        mangainfo.chapternames.add(v1.getattribute('title'))
      end
    end
    InvertStrings(mangainfo.chapterlinks,mangainfo.chapternames)
    return no_error
  else
    return net_problem
  end
end

function getpagenumber()
  task.pagenumber=0
  task.pagelinks.clear()
  if http.get(MaybeFillHost(module.rooturl,url)) then
    local x=TXQuery.Create(http.Document)
    local v=string.match(x.xpathstring('//script[contains(., "var data")]/text()'), 'var data =(.-);')
    TXQuery.Create(v).xpathstringall('json(*).fullimg()', task.pagelinks)
    return true
  else
    return false
  end
end

function getnameandlink()
  local s = module.rooturl..dirurl
  if url ~= '0' then s = s..'?offset='..tostring(tonumber(url) * perpage) end
  if http.get(s) then
    local x = TXQuery.Create(http.Document)
    x.XPathHREFAll('//*[@class="content_row"]//a[@class="title_link"]',links,names)
    return no_error
  else
    return net_problem
  end
end

function getdirectorypagenumber()
  page=1
  if http.GET(module.RootURL .. dirurl) then
    x = TXQuery.Create(http.Document)
    local s = tonumber(x.XPathString('//*[@id="pagination"]/a[last()]/substring-after(@href,"offset=")'))
    if s ~= '' then page=s end
    if page > 1 then page=math.floor(page / perpage) + 1 end
    return no_error
  else
    return net_problem
  end
end

function AddWebsiteModule(name, url, category)
  local m = NewModule()
  m.website = name
  m.rooturl = url
  m.category = category
  m.lastupdated='May 6, 2019'
  m.ongetinfo='getinfo'
  m.ongetpagenumber='getpagenumber'
  m.ongetnameandlink='getnameandlink'
  m.ongetdirectorypagenumber='getdirectorypagenumber'
  return m
end

function Init()
  local cat = 'Russian'
  AddWebsiteModule('MangaChanRU', 'http://manga-chan.me', cat)
  
  local cat = 'H-Sites'
  AddWebsiteModule('HentaiChanRU', 'http://h-chan.me', cat)
  AddWebsiteModule('YaoiChanRU', 'http://yaoi-chan.me', cat)
end
